base/libgit2/callbacks.jl

```diff
@@ -268,3 +253,52 @@ mirror_cb() = cfunction(mirror_callback, Cint, (Ptr{Ptr{Void}}, Ptr{Void}, Cstri
 credentials_cb() = cfunction(credentials_callback, Cint, (Ptr{Ptr{Void}}, Cstring, Cstring, Cuint, Ptr{Void}))
 "C function pointer for `fetchhead_foreach_callback`"
 fetchhead_foreach_cb() = cfunction(fetchhead_foreach_callback, Cint, (Cstring, Cstring, Ptr{GitHash}, Cuint, Ptr{Void}))
+
+"""
+Emulates the LibGit2 credential loop to allows testing of the credential_callback function
+without having to authenticate against a real server.
+"""
+function credential_loop(
+    valid_credential::AbstractCredential, url::AbstractString, user::AbstractString, allowed_types::UInt32,
+)
+    cb = credentials_cb()
+    libgitcred_ptr_ptr = Ref{Ptr{Void}}(C_NULL)
+    cache = CachedCredentials()
+    payload_ptr = Ref(Nullable{CachedCredentials}(cache))
+
+    # Number of times credentials were authenticated against. With the real LibGit2
+    # credential loop this would be how many times we sent credentials to the remote.
+    num_authentications = 0
+
+    # Emulate how LibGit2 uses the credential callback by repeatedly calling the function
+    # until we find valid credentials or an exception is raised.
+    err = Cint(0)
+    while err == 0
+        err = ccall(cb, Cint, (Ptr{Ptr{Void}}, Cstring, Cstring, Cuint, Ptr{Void}),
+            libgitcred_ptr_ptr, url, isempty(user) ? C_NULL : user, allowed_types, pointer_from_objref(payload_ptr))
+        num_authentications += 1
+
+        # Check if the callback provided us with valid credentials
+        if length(cache.cred) == 1 && first(values(cache.cred)) == valid_credential
+            break
+        end
+
+        if num_authentications > 50
+            error("Credential callback seems to be caught in an infinite loop")
+        end
+    end
+
+    return err, num_authentications
+end
+
+function credential_loop(
+    valid_credential::UserPasswordCredential;
+    url="https://github.com/test/package.jl", user="")
+    credential_loop(valid_credential, url, user, 0x000001)
```

Comment:
I usually give line continuations of block openers an extra level of indent to set them off without using up an extra vertical line, but we haven't totally standardized that.

Why is the default `url` set to anything at all here, that default value doesn't strike me as all that useful.